edtlib: add first class Binding object

mbolivar-nordic opened the pull request:
This is prep work for https://github.com/zephyrproject-rtos/zephyr/issues/28865.

The main goal is to make it possible to parse devicetree bindings without having a DTS file. This is necessary to generate documentation for available bindings as described in that issue.

This PR does not actually generate any documentation; the goal is just to get the library API changes reviewed and merged before proceeding with using them elsewhere.

mbolivar-nordic committed:
scripts: edtlib: tweak module docstring

Add a bit more information and clarification.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: edtlib: refactor for first class bindings

Add two new types: Binding and PropertySpec.

- Binding is a first-class abstraction for a devicetree binding
  file as defined by a YAML file in the Zephyr syntax.

- PropertySpec is a helper type which represents a property
  definition within a Binding.

Make the Binding constructor a new entry point to the library. This
enables users to deal with bindings as standalone entities, apart from
how they characterize a particular devicetree.

Rework the EDT and Node internals that deal with bindings as dict
values to use the Binding type instead. To make this less ambiguous,
use the variable name 'raw' when we're dealing with a binding as it's
parsed from YAML, and 'binding' when we're dealing with a Python
Binding object.

This commit drops support for the following legacy bindings keys

- '#cells': use '*-cells' instead (e.g. 'gpio-cells', 'pwm-cells')
- "child-bus: foo" and "child: bus: foo": use "bus:" instead
- "parent-bus" and "parent: bus: ": use "on-bus:" instead

Officially, legacy bindings have been gone since
6bf761fc0a2811b037 ("dts: Remove support for deprecated DTS binding
syntax"), so this is vestigial code, and I couldn't find any in-tree
users.

It also drops the convention in this file that ""-strings are
preferred.

I honestly don't understand why this was ever enforced; the file
itself admits single quotes are common in Python and we use them
elsewhere in Zephyr's Python code.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

galak reviewed (commented):
is it time to split edtlib.py into several files?  (not related, just a general question).

b0661 commented on scripts/dts/edtlib.py:
```diff
@@ -8,18 +8,20 @@
+In Zephyr, bindings are YAML files that describe devicetree nodes. Devicetree
+nodes are usually mapped to bindings via their 'compatible = "..."' property,
+but a binding can also come from a 'child-binding:' key in the binding for the
+parent devicetree node. See binding-template.yaml for details.
```
"See binding-template.yaml " makes binding-template.yaml kind of part of edtlib.py. As binding-template.yaml is not stored besides edtlib.py the full path would be more informative "[zephyr]/dts/bindings-template.yaml". For further improvement it could become part of the edtlib.py "library".

mbolivar-nordic replied:
> "See binding-template.yaml " makes binding-template.yaml kind of part of edtlib.py.

Yeah, I'll just drop this for now. I want to deal with that file in a separate PR if that's OK with you.

mbolivar-nordic commented on scripts/dts/edtlib.py:
```diff
@@ -292,18 +291,16 @@ def _init_graph(self):
         # Creates self._compat2binding. This is a dictionary that maps
-        # (<compatible>, <bus>) tuples (both strings) to (<binding>, <path>)
-        # tuples. <binding> is the binding in parsed PyYAML format, and <path>
-        # the path to the binding (nice for binding-related error messages).
+        # (<compatible>, <bus>) tuples (both strings) to Binding objects.
```
Will reword along these lines.

mbolivar-nordic commented on scripts/dts/edtlib.py:
```diff
@@ -8,18 +8,20 @@
-mapped to bindings via their 'compatible = "..."' property, but a binding can
-also come from a 'child-binding:' key in the binding for the parent devicetree
-node.
+In Zephyr, bindings are YAML files that describe devicetree nodes. Devicetree
```
> Why Zephyr, this is about Bindings in edtlib? -> Bindings are

Good point

mbolivar-nordic committed:
scripts: edtlib: cosmetic variable name change

Change "elm" to "elem" for readability.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
@b0661 comments addressed, thanks for keeping me honest on the zephyr-specific bits. I tacked on another patch for changing "elm" to "elem".

> is it time to split edtlib.py into several files? (not related, just a general question).

@galak I'm not ready for that yet, it's still navigable for me at least.

mbolivar-nordic commented:
>  dts: bindings: fix ti,lp503x.yaml child binding b157817

I had to prepend this to the series because the child-binding check in this refactor also makes sure that the child binding properties have types and descriptions as well. I think this is a valid restriction and arguably an oversight on edtlib's part prior to this.

mbolivar-nordic commented:
> dts: bindings: fix ti,lp503x.yaml child binding [b157817](https://github.com/zephyrproject-rtos/zephyr/commit/b157817e0d735b8d5fc9ae37d9a5e58c0f2e3cca)

@simonguinot fyi

galak commented:
@mbolivar-nordic a few questions:

1. do we think Binding and PropSpec can abstract away the actual binding format?  Wonder if we can have the same classes but populated from dt-schema bindings (would need additional features for parity)?

2. do you see adding some container class around collection of Bindings?  Not sure what exactly you need for the docs generation, but I had hacked something in the past to be able to generate DTS templates based on a binding.

mbolivar-nordic commented:
>     1. Wonder if we can have the same classes but populated from dt-schema bindings (would need additional features for parity)?

I had the same thought. Hopefully a later improvement! I'd need to study the dt-schema bindings.

>     2\. do you see adding some container class around collection of Bindings?

I was just gonna leave this up to the application if we don't have a concrete proposal right now.

galak commented:
> I was just gonna leave this up to the application if we don't have a concrete proposal right now.

Seems like something simple that given a set of binding dirs and optionally a set of compats would populate a list fo Bindings (or dict).  Looking at the old code I had, I think I was pulling `_init_compat2binding` into the Bindings class and have the constructor take `bindings_dirs` and optionally a list of `compats`

galak reviewed (changes requested):
LGTM.

There are a bunch of comments that reference _check_binding that need fixing.

mbolivar-nordic commented:
> There are a bunch of comments that reference _check_binding that need fixing.

Fixed, thanks

mbolivar-nordic commented:
ti,lp503x patch turned out to be a bug in my refactor. Fixed now.

mbolivar-nordic commented:
> > I was just gonna leave this up to the application if we don't have a concrete proposal right now.
> 
> Seems like something simple that given a set of binding dirs and optionally a set of compats would populate a list fo Bindings (or dict).  Looking at the old code I had, I think I was pulling `_init_compat2binding` into the Bindings class and have the constructor take `bindings_dirs` and optionally a list of `compats`

I'll add this in the docs generation PR if I can find a general purpose API.

galak committed:
scripts: edtlib: Remove deprecated "category" support

"category" support has been deprecated for over 2 releases.  Removing
support for the keyword.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts: edtlib: cleanup required arg to _check_prop_type_and_default

_check_prop_type_and_default doesn't use the required argument so remove
passing it to the function.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts: edt: Fix comment in _prop_val

_prop_val comment referred to one of the arguments called "optional"
however the code has changed to call that argument "required" now.  Fix
up the comment block to use the correct argument name and semantics of
that argument.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

mbolivar-nordic commented:
I added a few follow-on cleanups from @galak
